api/server/database/gorm/finding.go

```diff
@@ -170,36 +207,124 @@ func (s *FindingsTableHandler) UpdateFinding(finding types.Finding) (types.Findi
 		}
 	}
 
-	var dbFinding Finding
-	err := getExistingObjByID(s.DB, "Finding", *finding.Id, &dbFinding)
+	var dbObj Finding
+	err := getExistingObjByID(s.DB, findingSchemaName, *finding.Id, &dbObj)
 	if err != nil {
 		return types.Finding{}, err
 	}
 
-	dbFinding.Data, err = patchObject(dbFinding.Data, finding)
+	var dbFinding types.Finding
+	err = json.Unmarshal(dbObj.Data, &dbFinding)
 	if err != nil {
-		return types.Finding{}, fmt.Errorf("failed to apply patch: %w", err)
+		return types.Finding{}, fmt.Errorf("failed to convert DB model to API model: %w", err)
 	}
 
-	if err := s.DB.Save(&dbFinding).Error; err != nil {
-		return types.Finding{}, fmt.Errorf("failed to save finding in db: %w", err)
+	finding.Revision = bumpRevision(dbFinding.Revision)
+
+	dbObj.Data, err = patchObject(dbObj.Data, finding)
+	if err != nil {
+		return types.Finding{}, fmt.Errorf("failed to apply patch: %w", err)
 	}
 
-	// TODO(sambetts) Maybe this isn't required now because the DB isn't
-	// creating any of the data (like the ID) so we can just return the
-	// finding pre-marshal above.
-	var sc types.Finding
-	err = json.Unmarshal(dbFinding.Data, &sc)
+	var ret types.Finding
+	err = json.Unmarshal(dbObj.Data, &ret)
 	if err != nil {
 		return types.Finding{}, fmt.Errorf("failed to convert DB model to API model: %w", err)
 	}
-	return sc, nil
+
+	// Check uniqueness
+	existingFinding, err := s.checkUniqueness(ret)
+	if err != nil {
+		var conflictErr *common.ConflictError
+		if errors.As(err, &conflictErr) {
+			return *existingFinding, err
+		}
+		return types.Finding{}, fmt.Errorf("failed to check existing finding: %w", err)
+	}
+
+	if err := s.DB.Save(&dbObj).Error; err != nil {
+		return types.Finding{}, fmt.Errorf("failed to save finding in db: %w", err)
+	}
+
+	return ret, nil
 }
 
 func (s *FindingsTableHandler) DeleteFinding(findingID types.FindingID) error {
 	if err := deleteObjByID(s.DB, findingID, &Finding{}); err != nil {
-		return fmt.Errorf("failed to delete asset: %w", err)
+		return fmt.Errorf("failed to delete finding: %w", err)
 	}
 
 	return nil
 }
+
+//nolint:cyclop
+func (s *FindingsTableHandler) checkUniqueness(finding types.Finding) (*types.Finding, error) {
+	discriminator, err := finding.FindingInfo.ValueByDiscriminator()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get value by discriminator: %w", err)
+	}
+
+	if finding.Id == nil || *finding.Id == "" {
+		return &types.Finding{}, &common.BadRequestError{
+			Reason: "finding ID is required",
+		}
+	}
+
+	// Construct filter based on discriminator type
+	// Use info properties that make the finding unique, check package scanner/findingkey.
+	var filter string
+	switch info := discriminator.(type) {
+	case types.PackageFindingInfo:
+		key := findingkey.GeneratePackageKey(info)
+		filter = fmt.Sprintf("id ne '%s' and ", *finding.Id) + key.Filter()
```

Comment:
nit: filter seems to be the same for all cases and can be defined only once after switch